pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -544,16 +544,25 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 			return nil, fmt.Errorf("pod %q container %q changed request for resource %q from %d to %d", string(podUID), contName, resource, devices.Len(), required)
 		}
 	}
-	if needed == 0 {
-		// No change, no work.
-		return nil, nil
-	}
+
 	klog.V(3).InfoS("Need devices to allocate for pod", "deviceNumber", needed, "resourceName", resource, "podUID", string(podUID), "containerName", contName)
+	healthyDevices, hasRegistered := m.healthyDevices[resource]
+
 	// Check if resource registered with devicemanager
-	if _, ok := m.healthyDevices[resource]; !ok {
+	if !hasRegistered {
 		return nil, fmt.Errorf("can't allocate unregistered device %s", resource)
 	}
 
+	// Check if registered resource has healthy devices
+	if healthyDevices.Len() == 0 {
+		return nil, fmt.Errorf("can't allocate unhealthy devices %s", resource)
+	}
+
+	if needed == 0 {
+		// No change, no work.
+		return nil, nil
+	}
+
```

Comment:
As far as I can tell, this is the only "core" change (and it seems reasonable / logical).
The rest are changes to testing, is that correct?

Reply:
Actually, I have one question here -- your PR description mentioned that you wanted to check that all already-allocated devices were in fact healthy (and error out otherwise). I don't see this logic here. What you have in this PR is strictly better than what we had before, but I'm curious if this check was left out on purpose or overlooked.

Reply:
> As far as I can tell, this is the only "core" change (and it seems reasonable / logical).
The rest are changes to testing, is that correct?

Yes, this is accurate. Please note that https://github.com/kubernetes/kubernetes/pull/115107 would need to merge first as we need that for e2e tests implementation here. Also, due to some suggestions on that PR I have proposed a file based synchronization approach (as [3rd commit](https://github.com/kubernetes/kubernetes/pull/115107/commits/2aabf15c370ca727067f550a7ff225ec6d85af57) on that PR) diverging slightly sample device plugin changes in these two PRs temporarily. Once we reach consensus on that, I will rebase.

> Actually, I have one question here -- your PR description mentioned that you wanted to check that all already-allocated devices were in fact healthy (and error out otherwise). I don't see this logic here. What you have in this PR is strictly better than what we had before, but I'm curious if this check was left out on purpose or overlooked.

The change here: 
https://github.com/kubernetes/kubernetes/pull/114640/commits/6d80d27b8204b32b82da4e6ad3c10855327d36b0#diff-623f1164bd7d648ee0421742187c76b5071b44315aee695f29edc0203b6c762dR556-R559 is implementing that logic.
Previously, while iterating through `m.healthyDevices` we **only** checked if devices had registered which meant that it was okay to proceed forward with allocation if we had devices that previously registered even if they are unhealthy (and therefore unsuitable to be allocated after reboot either because the device plugin pod didn't started or didn't register itself yet to kubelet) .
Now, we obtain the list of `healthyDevices`, check its length to validate the presence of healthy devices and rule out `healthyDevices` with zero length as that is the case that corresponds to registered **unhealthy** devices. We only proceed to allocation if both these conditions are met ensuring that these devices are actually healthy and don't just appear to be because at some stage they had registered to kubelet.

Reply:
Maybe I misinterpreted what you meant then.

I thought you were saying you planned to validate that all of the already-allocated devices (i.e. those present in the `devices` variable at this point in the code) are healthy. The check you have in place just ensures that "at least one" device is healthy, but not that the full set of already-allocated devices are all healthy.

Reply:
I see what you mean, we can add a check to make sure that the `devices` retrieved from the checkpoint file are healthy as well.

I had been working under the assumption that nothing changes in the environment and once the device plugin has registered we would have enough devices to fulfill the pod resource requirement. I now see you have a valid point, we could have a case where a device plugin might have  re-registered but there aren't enough healthy devices e.g. a node had two NICs and devices are allocated from two NIC to a pod but one of them was removed before node reboot.

We can improve the current implementation by checking that there are at least `devices.Len()` number of devices instead of the current naive check of "at least one" healthy device. Also, we would have to check if `devices` is a subset of `healthyDevices`.

The only case that I want like to explicitly point out is where devices previously allocated (and thus retrieved from the checkpoint) are not healthy anymore but there are enough devices that can be allocated to the pod. In that case as well we would be returning an admission error. I think that is fair because without that we would proceed as normal but the pod would fail when it will try to access unhealthy devices "allocated" to it.

Let me know if that makes sense.

Reply:
Thought about this a bit more...
After node reboot, for all the registered devices  `m.healthyDevices` field is reset to zero because of which device capacity is 0. This is intentional as we want to prevent additional pods requesting devices to be scheduled on the node.  

Since `devices` is not empty, we check if additional resources need to be allocated`needed = needed - devices.Len()`

it is sufficient to check if healthyDevices.len is zero as that is an indication of the device plugin not being registered.

Since `healthyDevices` is empty we can't cross check if the devices retrieved are healthy or not.

Reply:
I was talking about the state of things after ensuring that healthyDevices.Len() > 0 (and consequently before checking if needed == 0).

The set of already-allocated-devices at this point are just assumed to all be healthy if we don’t explicitly check for this. 

I’m just wondering if we want to cross check them against the set of devices in healthyDevices and if they are not all there, we error out.

Reply:
I see, I will handle that case as well. Thanks.